chirp/drivers/hf90.py

```diff
@@ -0,0 +1,567 @@
+
+        rs = RadioSetting("type", "Type",
+                          RadioSettingValueList(
+                              self.HF90_TYPES,
+                              self.HF90_TYPES[_settings.type]))
+        rs.set_apply_callback(self.apply_type, _settings)
+        grp.append(rs)
+
+        if isinstance(self, EarlyHF90Radio):
+            rs = RadioSetting("selcall_id", "Selcall ID (0 for standard type)",
+                              RadioSettingValueInteger(
+                                  0, 9999,
+                                  self._get_selcall_id(_settings)))
+            rs.set_apply_callback(self.apply_selcall_id, _settings)
+            grp.append(rs)
+        else:
+            rs = RadioSetting("selcall_id", "Selcall ID",
+                              RadioSettingValueInteger(
+                                  1, 9999,
+                                  self._get_selcall_id(_settings)))
+            rs.set_apply_callback(self.apply_selcall_id, _settings)
+            grp.append(rs)
+
+        rs = RadioSetting("f_p_program", "Front panel programming",
+                          RadioSettingValueBoolean(_settings.f_p_program))
+        grp.append(rs)
+
+        rs = RadioSetting("auto_tune", "Auto tuning",
+                          RadioSettingValueBoolean(_settings.auto_tune))
+        grp.append(rs)
+
+        rs = RadioSetting("mode_toggle", "Mode toggling",
+                          RadioSettingValueBoolean(_settings.mode_toggle))
+        grp.append(rs)
+
+        hp = int(_settings.tx_power == 3)
+        rs = RadioSetting("tx_power", "Tx Power",
+                          RadioSettingValueList(self.HF90_POWER_LEVELS,
+                                                self.HF90_POWER_LEVELS[hp]))
+        rs.set_apply_callback(self.apply_txpower, _settings)
+        grp.append(rs)
+
+        return RadioSettings(grp)
+
+    def set_settings(self, uisettings):
+        _settings = self._memobj.settings
+
+        for element in uisettings:
+            if not isinstance(element, RadioSetting):
+                self.set_settings(element)
+                continue
+            if not element.changed():
+                continue
+
+            try:
+                name = element.get_name()
+                value = element.value
+
+                if element.has_apply_callback():
+                    LOG.debug("Using apply callback")
+                    element.run_apply_callback()
+                else:
+                    setattr(_settings, name, value)
+
+                LOG.debug("Setting %s: %s" % (name, value))
+            except Exception as e:
+                LOG.debug(element.get_name())
+                raise
+
+    def get_memory(self, number):
+        _mem = self._memobj.memory[number - 1]
+
+        mem = chirp_common.Memory()
+        mem.number = number
+
+        _rx = _mem.rx0_7 | (_mem.rx8_15 << 8) | (_mem.rx16_19 << 16)
+        _tx = _mem.tx0_3 | (_mem.tx4_7 << 4) | (_mem.tx8_11 << 8)
```

Comment:
Why are you doing this? Aren't these just an `le24` type?

Reply:
Okay, but `_rx` is really a `ul24` right? I read the first and didn't pay attention to the fact that the tx one was (weirdly) not the same.

Reply:
Sorry, I misspoke. Tx and rx are 20 bit quantities (10 nybbles altogether).

Reply:
You're right, there is a simpler representation.
`  ul16 rx0_15;
  u8 tx0_3:4, rx16_19:4;
  ul16 tx4_19;
`

Reply:
Sorry, I see now that the tx and rx bytes are interleaved with each other in memory. I just wasn't seeing it before. I dunno why someone would design something like that, but there are similarly crazy things in other radios, so it must be in the water at conferences or something.